### PR TITLE
fix speed limit restriction

### LINKED ISSUE #6407: Speed limit edition and usage with no value fails
### What happened?

When we create a speed limit with no default value `0` is given to editoast.
Core behavior is weird having a speed limit of `0` (it uses the rolling stock speed limit).

### What did you expect to happen?

- [x] Front should provide a `null` speed limit when not provided (allowed by the model) => #6428
- [x] Editoast should expect speed greater than `0` => #6436
- [x] Core should expect speed greater than `0` => #6436
- [x] We could provide a migration to transform all `0` speed to `null` => #6436

### How can we reproduce it (as minimally and precisely as possible)?

1. Use this [infra](https://github.com/osrd-project/osrd/files/13992351/Broken.Infra.id5.railjson.json)
2. Create a train (left to right)
3. Its MRSP should be `80km/h`

### What operating system, browser and environment are you using?

- **Browser**: Firefox v109
- **OS**: Nixos v22.11
- **Env**: Local

### OSRD version (top right corner `Account` button > `Informations`)

572a549e28a1e551b873cef3c9d4e7888e03654b

### PR DESCRIPTION
**Pull request by kmer2016**
Closes #6407 
- [x] Editoast should return an error when the speed is less than 0
  - [x] add validation when Deserialize SpeedSection
  - [x] add check on infra endpoints (edition and loading)
- [x] Core should return an error when the speed is less than 0
- [x] migration to transform all 0 speed to null

### CODE REVIEW

File: editoast/src/schema/speed_section.rs [hunk 1]
```diff
@@ -15,21 +15,44 @@ use derivative::Derivative;
...
+    #[validate(custom = "validate_speed_limit")]
     pub speed_limit: Option<f64>,
+    #[validate(custom = "validate_speed_limit_by_tag")]
     pub speed_limit_by_tag: HashMap<NonBlankString, f64>,
     pub track_ranges: Vec<ApplicableDirectionsTrackRange>,
     #[serde(default)]
     pub extensions: SpeedSectionExtensions,
 }
 
+pub const SPEED_LIMIT_ERROR: &str = "speed_limit must be greater than 0";
```
Comment (flomonster) [inline editoast/src/schema/speed_section.rs]: It looks like this is only useful for testing purposes. We could only check that the validation fails and inline this (keeping it private).

File: editoast/src/views/infra/railjson.rs [hunk 2]
```diff
@@ -161,6 +171,17 @@ async fn post_railjson(
...
+fn invalid_speed_sections(speed_sections: &[SpeedSection]) -> Vec<String> {
+    let mut errors = Vec::new();
+    for section in speed_sections {
+        if let Err(_e) = section.validate() {
+            let error_message = format!("Validation error on speedsection {}", section.get_id());
+            errors.push(error_message);
+        }
+    }
+    errors
+}
```
Comment (flomonster) [inline editoast/src/views/infra/railjson.rs]: We should implement the validation trait to `Railjson`. Doing so could allow us to simply do something like:

```rust
let railjson = railjson.into_inner().validate()?;
```

File: editoast/src/views/infra/mod.rs [hunk 3]
```diff
@@ -101,6 +101,9 @@ pub enum InfraApiError {
     #[error("Infra '{infra_id}', could not be found")]
     #[editoast_error(status = 404)]
     NotFound { infra_id: i64 },
+    #[error("Validation errors on speed section(s) : '{}'", .speed_sections.join(", "))]
+    #[editoast_error(status = 422)]
+    SpeedSectionValidationError { speed_sections: Vec<String> },
```
Comment (flomonster) [inline editoast/src/views/infra/mod.rs]: We don't need this. Validation errors can be transformed automatically into `EditoastError` (check `errors.rs`)

File: editoast/src/schema/operation/update.rs [hunk 4]
```diff
@@ -94,6 +95,13 @@ impl DataObject {
         if obj_railjson.get_id() != &update.obj_id {
             return Err(OperationError::ModifyId.into());
         }
+
+        if let RailjsonObject::SpeedSection { railjson } = &obj_railjson {
+            railjson
+                .validate()
+                .map_err(|err| OperationError::InvalidPatch(err.to_string()))?;
+        };
```
Comment (flomonster) [inline editoast/src/schema/operation/update.rs]: Same here, we can implement `Validate` to the `RailjsonObject` struct and `Operation`.

File: editoast/migrations/2024-01-26-150237_change_zero_speed_limit_to_null/up.sql [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    COALESCE(
+        (
+            SELECT jsonb_object_agg(key, value)
+            FROM jsonb_each(data->'speed_limit_by_tag')
+            WHERE (value::float > 0)
+        ),
+        '{}'::jsonb
+    )
+)
+WHERE data->'speed_limit_by_tag' IS NOT NULL;
```
Comment (flomonster) [inline editoast/migrations/2024-01-26-150237_change_zero_speed_limit_to_null/up.sql]: Missing end of line

File: core/src/test/java/fr/sncf/osrd/envelope_sim_infra/MRSPTest.java [hunk 6]
```diff
@@ -50,7 +50,7 @@ public void setUp() throws IOException, URISyntaxException {
         var speedSection1 = new RJSSpeedSection(NAME, SPEED1, Map.of(TRAIN_TAG1, SPEED2),
                 List.of(new RJSApplicableDirectionsTrackRange("TA0", ApplicableDirection.BOTH,
                         POSITION1, POSITION2)));
-        var speedSection2 = new RJSSpeedSection(NAME, 0, Map.of(TRAIN_TAG2, SPEED3),
+        var speedSection2 = new RJSSpeedSection(NAME, Double.POSITIVE_INFINITY, Map.of(TRAIN_TAG2, SPEED3),
```
Comment (flomonster) [inline core/src/test/java/fr/sncf/osrd/envelope_sim_infra/MRSPTest.java]:
```suggestion
        var speedSection2 = new RJSSpeedSection(NAME, null, Map.of(TRAIN_TAG2, SPEED3),
```

**Review by flomonster — COMMENTED** [review]
I reviewed the editoast part. It seems like the initial approach of using `validator::Validate` will be heavy. My comments are aligned with this approach.

We can discuss implementing manually `Deserialize` instead of `Validate`.

File: core/src/test/java/fr/sncf/osrd/envelope_sim_infra/MRSPTest.java (hunk 6 above)
Comment (kmer2016) [reply core/src/test/java/fr/sncf/osrd/envelope_sim_infra/MRSPTest.java]: RJSSpeedSection expect a `double` so I can't instanciate with `null`, Do I change the definition to use `Double` instead ?

File: editoast/src/schema/railjson.rs [hunk 7]
```diff
@@ -14,8 +14,9 @@ use diesel_async::{AsyncPgConnection as PgConnection, RunQueryDsl};
 use editoast_derive::EditoastError;
 use serde::{de::DeserializeOwned, Deserialize, Serialize};
 use thiserror::Error;
+use validator::Validate;
 
-#[derive(Deserialize, Derivative, Serialize, Clone, Debug)]
+#[derive(Deserialize, Derivative, Serialize, Clone, Debug, Validate)]
```
Comment (flomonster) [inline editoast/src/schema/railjson.rs]:
```suggestion
#[derive(Deserialize, Derivative, Serialize, Clone, Debug)]
```

**Review by flomonster — APPROVED** [review]
A small fix and we're good to merge it!

**Review by Math-R — APPROVED** [review]
LGTM =)

File: core/src/main/java/fr/sncf/osrd/infra/api/tracks/undirected/SpeedLimits.java [hunk 8]
```diff
@@ -2,7 +2,10 @@
 
 import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.Sets;
+import com.squareup.moshi.JsonDataException;
 import fr.sncf.osrd.railjson.schema.infra.trackranges.RJSSpeedSection;
+import fr.sncf.osrd.reporting.exceptions.ErrorType;
```
Comment (eckter) [inline core/src/main/java/fr/sncf/osrd/infra/api/tracks/undirected/SpeedLimits.java]:
```suggestion
```

**Review by eckter — APPROVED** [review]
LGTM

**Pull request merged**